src/gpu/metal/SDL_gpu_metal.m

```diff
@@ -0,0 +1,3999 @@
+        for (Uint32 i = 0; i < computePipeline->uniformBufferCount; i += 1) {
+            [commandBuffer->computeEncoder
+                setBuffer:commandBuffer->computeUniformBuffers[i]->handle
+                   offset:commandBuffer->computeUniformBuffers[i]->drawOffset
+                  atIndex:i];
+        }
+
+        commandBuffer->needComputeUniformBind = SDL_FALSE;
+    }
+}
+
+static void METAL_DrawIndexedPrimitives(
+    SDL_GpuCommandBuffer *commandBuffer,
+    Uint32 baseVertex,
+    Uint32 startIndex,
+    Uint32 primitiveCount,
+    Uint32 instanceCount)
+{
+    MetalCommandBuffer *metalCommandBuffer = (MetalCommandBuffer *)commandBuffer;
+    SDL_GpuPrimitiveType primitiveType = metalCommandBuffer->graphicsPipeline->primitiveType;
+    Uint32 indexSize = IndexSize(metalCommandBuffer->indexElementSize);
+
+    METAL_INTERNAL_BindGraphicsResources(metalCommandBuffer);
+
+    [metalCommandBuffer->renderEncoder
+        drawIndexedPrimitives:SDLToMetal_PrimitiveType[primitiveType]
+                   indexCount:PrimitiveVerts(primitiveType, primitiveCount)
+                    indexType:SDLToMetal_IndexType[metalCommandBuffer->indexElementSize]
+                  indexBuffer:metalCommandBuffer->indexBuffer->handle
+            indexBufferOffset:metalCommandBuffer->indexBufferOffset + (startIndex * indexSize)
+                instanceCount:instanceCount
+                   baseVertex:baseVertex
+                 baseInstance:0];
+}
+
+static void METAL_DrawPrimitives(
+    SDL_GpuCommandBuffer *commandBuffer,
+    Uint32 vertexStart,
+    Uint32 primitiveCount)
+{
+    MetalCommandBuffer *metalCommandBuffer = (MetalCommandBuffer *)commandBuffer;
+    SDL_GpuPrimitiveType primitiveType = metalCommandBuffer->graphicsPipeline->primitiveType;
+
+    METAL_INTERNAL_BindGraphicsResources(metalCommandBuffer);
+
+    [metalCommandBuffer->renderEncoder
+        drawPrimitives:SDLToMetal_PrimitiveType[primitiveType]
+           vertexStart:vertexStart
+           vertexCount:PrimitiveVerts(primitiveType, primitiveCount)];
+}
+
+static void METAL_DrawPrimitivesIndirect(
+    SDL_GpuCommandBuffer *commandBuffer,
+    SDL_GpuBuffer *buffer,
+    Uint32 offsetInBytes,
+    Uint32 drawCount,
+    Uint32 stride)
+{
+    MetalCommandBuffer *metalCommandBuffer = (MetalCommandBuffer *)commandBuffer;
+    MetalBuffer *metalBuffer = ((MetalBufferContainer *)buffer)->activeBuffer;
+    SDL_GpuPrimitiveType primitiveType = metalCommandBuffer->graphicsPipeline->primitiveType;
+
+    METAL_INTERNAL_BindGraphicsResources(metalCommandBuffer);
+
+    /* Metal: "We have multi-draw at home!"
+     * Multi-draw at home:
+     */
+    for (Uint32 i = 0; i < drawCount; i += 1) {
+        [metalCommandBuffer->renderEncoder
+                  drawPrimitives:SDLToMetal_PrimitiveType[primitiveType]
+                  indirectBuffer:metalBuffer->handle
+            indirectBufferOffset:offsetInBytes + (stride * i)];
+    }
```

Comment:
Thanks for the thorough explanation. I think adding support for indirect command buffers is a bigger can of worms than we would like to take on at the moment, since it would require a substantial rework of the backend internals. The second method you mention sounds like something we could look into in the future as a potential optimization though!